### ISSUE TITLE
Correct way to train a new Orca? (with access to a cluster)

### ISSUE THREAD

**Issue by adney11** [Soheil-ab/Orca#25]
Hi @Soheil-ab,

I had a few questions regarding training Orca.

The paper mentions you use 256 actors interacting with different environments - can you please give some insight on:

1. Does each actor get one trace over which it will collect 50k samples - after which the actor dies, and the learner waits for all actors to die? Does this mean your training dataset had a total of 256 generated traces?
2. How did you distributed these actors across your pool of servers - could you please share your workflow or some resource to go over for this? Did you have a single coordinator distributing all the actor tasks, or did you have to manually start a learner on one machine, and then signal the other machines to start their actor counterpart?
3. If you could share your trace generator, or methodology of trace generation, that would be very helpful too.
4. How can we be confident that our newly trained Orca has been trained correctly?
5. I see "/cpu" being used in the given code, but the paper mentions your training cluster includes a GPU - did you use a GPU to train, or is it not very relevant since the networks used for Orca are not very big?

Thank you for taking the time to read and answer these questions!

**Comment by Soheil-ab** [comment]
Hi @adney11 

0) Number of actors was more than 256 

1) As described in the paper, each actor deals with an environment, collects the required samples, and dies. When all actors are done, one round of training is finished. As you may see, the total number of (concurrent) actors is a function of your available resources. In our scenario, we had more environment than the number of actors. That means for "fast" actors, we assigned new environments to them ...

2) There are multiple different ways to deal with this. For the model trained in the SIGCOMM'20 paper, the process was like this: I had an orchestrator responsible for bringing up the entire training ...
In particular, the orchestrator: 
 2-0) starts the learner, sets up the replay memory and corresponding modules
 2-1) assigns diff. environments to diff. servers
 2-2) maps env. to actors on selected servers
 2-3) monitors the behavior of actors during the process ...
 2-4) assigns more new environments to "fast" actors if needed

3) trace generation is quite straightforward. You just need to know how your emulator interprets traces. Then, you can either generate traces offline, make the pool, and use them by the orchestrator later. Or, let the actors generate them on demand. I used "the pool" approach ... 

We used Mahimahi to control not just the trace but other network parameters like bottleneck queue size. The task of executing an environment with given parameters was given to each actor. 

4) That's a nice question! The easiest way to make sure that you have trained an Orca is to make sure that the "Return vs time/step/epoch" graph is converged (to a good/high value). There are other ways like checking the loss graph. But, I personally prefer the return graph (Figs 4 and 5 on SIGCOMM'20 paper). How to monitor that? We have already put some useful functions in the released code that you can use. It's as simple as bringing up the Tensorboard and visualizing the graphs during the training (You can also use other tricks like having a function to kill the training after no meaningful improvements happens ...)

- In the configuration file, you'll see a parameter called "eval_frequency". This was used to generate the return graph. Knowing that may help you read the code faster. You can check the code for more details. 

5) We trained using GPUs. The code provides you with a simple config param called "device" that you can use for that purpose. 

Hope that helps,
Soheil

**Comment by adney11** [comment]
Thank you @Soheil-ab for the detailed explanations! Just a few more doubts, I appreciate your patience!

"2-3) monitors the behavior of actors during the process ...
2-4) assigns more new environments to "fast" actors if needed"
Would it be possible to share some reference for your training orchestrator? 

It's not clear how to assign new environments, without killing off the actor's d5 component (i.e, if we were to use the provided actor.sh script as the distributed actor process, which just calls the orca-server executable - which indirectly starts the d5 actor component, does the sending of data, and then dies after the set number of samples are collected).

Is it right to assume that you let your actor d5 processes remain running on your remote servers, and just signal the "fast" actor to run a new client with a new environment inside mahimahi, for the originally started d5 process?
